File: app/src/main/java/closer/vlllage/com/closer/handler/group/GroupHandler.java
package closer.vlllage.com.closer.handler.group;

import android.support.annotation.NonNull;
import android.view.View;
import android.widget.ImageView;
import android.widget.TextView;

import org.greenrobot.essentials.StringUtils;

import java.util.ArrayList;
import java.util.List;

import closer.vlllage.com.closer.R;
import closer.vlllage.com.closer.api.models.GroupResult;
import closer.vlllage.com.closer.handler.FeatureHandler;
import closer.vlllage.com.closer.handler.FeatureType;
import closer.vlllage.com.closer.handler.data.ApiHandler;
import closer.vlllage.com.closer.handler.data.DataHandler;
import closer.vlllage.com.closer.handler.data.PersistenceHandler;
import closer.vlllage.com.closer.handler.data.RefreshHandler;
import closer.vlllage.com.closer.handler.helpers.DefaultAlerts;
import closer.vlllage.com.closer.handler.helpers.DisposableHandler;
import closer.vlllage.com.closer.handler.helpers.PhotoLoader;
import closer.vlllage.com.closer.handler.helpers.ResourcesHandler;
import closer.vlllage.com.closer.handler.helpers.Val;
import closer.vlllage.com.closer.handler.phone.NameHandler;
import closer.vlllage.com.closer.pool.PoolMember;
import closer.vlllage.com.closer.store.StoreHandler;
import closer.vlllage.com.closer.store.models.Event;
import closer.vlllage.com.closer.store.models.Group;
import closer.vlllage.com.closer.store.models.GroupContact;
import closer.vlllage.com.closer.store.models.GroupContact_;
import closer.vlllage.com.closer.store.models.GroupInvite;
import closer.vlllage.com.closer.store.models.GroupInvite_;
import closer.vlllage.com.closer.store.models.Group_;
import closer.vlllage.com.closer.store.models.Phone;
import io.objectbox.android.AndroidScheduler;
import io.objectbox.reactive.DataSubscription;
import io.reactivex.subjects.BehaviorSubject;
import io.reactivex.subjects.PublishSubject;

public class GroupHandler extends PoolMember {

    private TextView groupName;
    private TextView groupAbout;
    private ImageView backgroundPhoto;
    private TextView peopleInGroup;
    private View settingsButton;
    private Group group;
    private GroupContact groupContact;
    private BehaviorSubject<Group> groupChanged = BehaviorSubject.create();
    private PublishSubject<Group> groupUpdated = PublishSubject.create();
    private BehaviorSubject<Event> eventChanged = BehaviorSubject.create();
    private BehaviorSubject<Phone> phoneChanged = BehaviorSubject.create();
    private List<String> contactNames = new ArrayList<>();
    private List<String> contactInvites = new ArrayList<>();
    private DataSubscription groupDataSubscription;

    public void attach(TextView groupName, ImageView backgroundPhoto, TextView groupAbout, TextView peopleInGroup, View settingsButton) {
        this.groupName = groupName;
        this.backgroundPhoto = backgroundPhoto;
        this.groupAbout = groupAbout;
        this.peopleInGroup = peopleInGroup;
        this.settingsButton = settingsButton;
    }

    public void setGroupById(String groupId) {
        if (groupId == null) {
            setGroup(null);
            return;
        }

        setGroup($(StoreHandler.class).getStore().box(Group.class).query()
                .equal(Group_.id, groupId)
                .build().findFirst());

        if (group == null) {
            $(DisposableHandler.class).add($(ApiHandler.class).getGroup(groupId)
                    .map(GroupResult::from)
                    .subscribe(group -> {
                        $(RefreshHandler.class).refresh(group);
                        setGroup(group);
                    }));
        }
    }

    private void onGroupSet(@NonNull Group group) {
        setGroupContact();
        peopleInGroup.setText("");

        if ($(Val.class).isEmpty(group.getAbout())) {
            groupAbout.setVisibility(View.GONE);
        } else {
            groupAbout.setVisibility(View.VISIBLE);
            groupAbout.setText(group.getAbout());
        }

        $(DisposableHandler.class).add($(StoreHandler.class).getStore().box(GroupContact.class).query()
                .equal(GroupContact_.groupId, group.getId())
                .build()
                .subscribe()
                .on(AndroidScheduler.mainThread())
                .observer(groupContacts -> {
                    $(GroupContactsHandler.class).setCurrentGroupContacts(groupContacts);
                    contactNames = new ArrayList<>();
                    for (GroupContact groupContact : groupContacts) {
                        contactNames.add($(NameHandler.class).getName(groupContact));
                    }

                    redrawContacts();
                }));

        $(DisposableHandler.class).add($(StoreHandler.class).getStore().box(GroupInvite.class).query()
                .equal(GroupInvite_.group, group.getId())
                .build().subscribe().on(AndroidScheduler.mainThread())
                .observer(groupInvites -> {
                    contactInvites = new ArrayList<>();
                    for (GroupInvite groupInvite : groupInvites) {
                        contactInvites.add($(ResourcesHandler.class).getResources().getString(R.string.contact_invited_inline, $(NameHandler.class).getName(groupInvite)));
                    }
                    redrawContacts();
                }));

        if ($(FeatureHandler.class).has(FeatureType.FEATURE_MANAGE_PUBLIC_GROUP_SETTINGS)) {
            settingsButton.setVisibility(View.VISIBLE);
        }
    }

    private void setGroupContact() {
        if ($(PersistenceHandler.class).getPhoneId() == null) {
            return;
        }

        groupContact = $(StoreHandler.class).getStore().box(GroupContact.class).query()
                .equal(GroupContact_.groupId, group.getId())
                .equal(GroupContact_.contactId, $(PersistenceHandler.class).getPhoneId())
                .build().findFirst();
    }

    private void redrawContacts() {
        List<String> names = new ArrayList<>();
        names.addAll(contactNames);
        names.addAll(contactInvites);

        if (names.isEmpty()) {
            peopleInGroup.setVisibility(View.GONE);
            peopleInGroup.setText(R.string.add_contact);
            return;
        }

        peopleInGroup.setVisibility(View.VISIBLE);

        peopleInGroup.setText(StringUtils.join(names, ", "));
    }

    private void setGroup(Group group) {
        this.group = group;

        if (groupDataSubscription != null) {
            $(DisposableHandler.class).dispose(groupDataSubscription);
        }

        if (group != null) {
            onGroupSet(group);
            groupChanged.onNext(group);
            setEventById(group.getEventId());
            setPhoneById(group.getPhoneId());
            $(RefreshHandler.class).refreshGroupMessages(group.getId());
            $(RefreshHandler.class).refreshGroupContacts(group.getId());

            groupDataSubscription = $(StoreHandler.class).getStore().box(Group.class).query()
                    .equal(Group_.id, group.getId())
                    .build()
                    .subscribe()
                    .onlyChanges()
                    .on(AndroidScheduler.mainThread())
                    .observer(groups -> {
                        if (groups.isEmpty()) return;
                        redrawContacts();
                        groupUpdated.onNext(groups.get(0));
                        $(RefreshHandler.class).refreshGroupContacts(group.getId());
                    });

            $(DisposableHandler.class).add(groupDataSubscription);
        }

        showGroupName(group);
    }

    public void showGroupName(Group group) {
        if (group == null) {
            groupName.setText(R.string.not_found);
            return;
        }

        if (group.hasPhone()) {
            $(DisposableHandler.class).add($(DataHandler.class).getPhone(group.getPhoneId()).subscribe(
                    phone -> {
                        groupName.setText(phone.getName());
                    }, error -> $(DefaultAlerts.class).thatDidntWork()
            ));
        } else {
            groupName.setText($(Val.class).of(group.getName(), $(ResourcesHandler.class).getResources().getString(R.string.app_name)));
        }
    }

    public void setGroupBackground(Group group) {
        if (group.getPhoto() != null) {
            backgroundPhoto.setVisibility(View.VISIBLE);
            backgroundPhoto.setImageDrawable(null);
            $(PhotoLoader.class).softLoad(group.getPhoto(), backgroundPhoto);
        } else {
            backgroundPhoto.setVisibility(View.GONE);
        }
    }

    private void setEventById(String eventId) {
        if (eventId == null) {
            return;
        }

        $(DisposableHandler.class).add($(DataHandler.class).getEventById(eventId)
                .subscribe(event -> eventChanged.onNext(event),
                        error -> $(DefaultAlerts.class).thatDidntWork()));
    }

    private void setPhoneById(String phoneId) {
        if (phoneId == null) {
            return;
        }

        $(DisposableHandler.class).add($(DataHandler.class).getPhone(phoneId)
                .subscribe(phone -> phoneChanged.onNext(phone),
                        error -> $(DefaultAlerts.class).thatDidntWork()));
    }

    public Group getGroup() {
        return group;
    }

    public GroupContact getGroupContact() {
        return groupContact;
    }

    public BehaviorSubject<Group> onGroupChanged() {
        return groupChanged;
    }

    public PublishSubject<Group> onGroupUpdated() {
        return groupUpdated;
    }

    public BehaviorSubject<Event> onEventChanged() {
        return eventChanged;
    }

    public BehaviorSubject<Phone> onPhoneChanged() {
        return phoneChanged;
    }
}

File: app/src/main/java/closer/vlllage/com/closer/handler/group/GroupHandler.kt
package closer.vlllage.com.closer.handler.group

import closer.vlllage.com.closer.R
import closer.vlllage.com.closer.handler.data.DataHandler
import closer.vlllage.com.closer.handler.data.PersistenceHandler
import closer.vlllage.com.closer.handler.data.RefreshHandler
import closer.vlllage.com.closer.handler.helpers.*
import closer.vlllage.com.closer.handler.phone.NameHandler
import closer.vlllage.com.closer.store.StoreHandler
import closer.vlllage.com.closer.store.models.*
import com.queatz.on.On
import io.objectbox.android.AndroidScheduler
import io.reactivex.Observable
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.subjects.BehaviorSubject
import io.reactivex.subjects.PublishSubject

class GroupHandler constructor(private val on: On) {

    private val connectionError = { throwable: Throwable ->
        throwable.printStackTrace()
        on<ConnectionErrorHandler>().notifyConnectionError()
    }

    private val disposableGroup = on<DisposableHandler>().group()

    var group: Group? = null
        set(group) {
            if (group?.id == field?.id) {
                return
            }

            field = group
            disposableGroup.clear()

            if (group == null) {
                return
            }

            onGroupSet(group)
            groupChanged.onNext(group)
            setEventById(group.eventId)
            setPhoneById(group.phoneId)
            on<RefreshHandler>().refreshGroupMessages(group.id!!)
            on<RefreshHandler>().refreshGroupContacts(group.id!!)

            disposableGroup.add(on<StoreHandler>().store.box(Group::class).query()
                    .equal(Group_.id, group.id!!)
                    .build()
                    .subscribe()
                    .onlyChanges()
                    .on(AndroidScheduler.mainThread())
                    .observer { groups ->
                        if (groups.isEmpty()) return@observer
                        groupUpdated.onNext(groups[0])
                        on<RefreshHandler>().refreshGroupContacts(group.id!!)
                    })

            on<PersistenceHandler>().phoneId?.let { phoneId ->
                disposableGroup.add(on<StoreHandler>().store.box(GroupMember::class).query()
                        .equal(GroupMember_.group, group.id!!)
                        .equal(GroupMember_.phone, phoneId)
                        .build()
                        .subscribe()
                        .on(AndroidScheduler.mainThread())
                        .observer { groupMembers ->
                            groupMemberChanged.onNext(if (groupMembers.isEmpty()) GroupMember() else groupMembers[0])
                        })
            }
        }

    var event: Event? = null
        get() = eventChanged.value
        private set

    var phone: Phone? = null
        get() = phoneChanged.value
        private set

    var groupContact: GroupContact? = null
        private set

    private val groupChanged = BehaviorSubject.create<Group>()
    private val groupUpdated = PublishSubject.create<Group>()
    private val groupNotFound = PublishSubject.create<Unit>()
    private val eventChanged = BehaviorSubject.create<Event>()
    private val phoneChanged = BehaviorSubject.create<Phone>()
    private val phoneUpdated = BehaviorSubject.create<Phone>()
    private val contactInfoChanged = BehaviorSubject.create<ContactInfo>()
    private val groupMemberChanged = BehaviorSubject.create<GroupMember>()
    private val contactInfo = ContactInfo()

    fun setGroupById(groupId: String?) {
        if (groupId == null) {
            group = null
            return
        }

        disposableGroup.add(on<DataHandler>().getGroup(groupId).observeOn(AndroidSchedulers.mainThread()).subscribe({
            group = it

            on<RefreshHandler>().refreshGroup(groupId)
        }, { on<DefaultAlerts>().message(on<ResourcesHandler>().resources.getString(R.string.group_not_found), on<ResourcesHandler>().resources.getString(R.string.group_not_found_description)) {
            groupNotFound.onNext(Unit)
        } }))
    }

    private fun onGroupSet(group: Group) {
        if (group.direct) {
            contactInfoChanged.onNext(contactInfo.also {
                it.contactNames.clear()
                it.contactInvites.clear()
            })
        }

        disposableGroup.add(on<StoreHandler>().store.box(GroupContact::class)
                .query()
                .equal(GroupContact_.groupId, group.id!!)
                .build()
                .subscribe()
                .on(AndroidScheduler.mainThread())
                .observer { groupContacts ->
                    groupContact = groupContacts.firstOrNull { it.contactId == on<PersistenceHandler>().phoneId!! }

                    on<GroupContactsHandler>().setCurrentGroupContacts(groupContacts)

                    if (!group.direct) {
                        contactInfo.contactNames.clear()
                        for (groupContact in groupContacts) {
                            contactInfo.contactNames.add(on<NameHandler>().getName(groupContact))
                        }

                        contactInfoChanged.onNext(contactInfo)
                    }
                })

        disposableGroup.add(on<StoreHandler>().store.box(GroupInvite::class)
                .query()
                .equal(GroupInvite_.group, group.id!!)
                .build()
                .subscribe()
                .on(AndroidScheduler.mainThread())
                .observer { groupInvites ->
                    if (!group.direct) {
                        contactInfo.contactInvites.clear()
                        for (groupInvite in groupInvites) {
                            contactInfo.contactInvites.add(on<ResourcesHandler>().resources.getString(R.string.contact_invited_inline, on<NameHandler>().getName(groupInvite)))
                        }
                        contactInfoChanged.onNext(contactInfo)
                    }
                })
    }

    private fun setEventById(eventId: String?) {
        if (eventId == null) {
            return
        }

        disposableGroup.add(on<DataHandler>().getEvent(eventId)
                .subscribe({ event -> eventChanged.onNext(event) },
                        { on<DefaultAlerts>().thatDidntWork() }))
    }

    private fun setPhoneById(phoneId: String?) {
        if (phoneId == null) {
            return
        }

        on<RefreshHandler>().refreshPhone(phoneId)

        disposableGroup.add(on<StoreHandler>().store.box(Phone::class).query()
                .equal(Phone_.id, phoneId)
                .build()
                .subscribe()
                .on(AndroidScheduler.mainThread())
                .observer {
                    if (it.isEmpty()) return@observer

                    val phone = it.first()

                    if (phoneChanged.value?.id != phone?.id) {
                        phoneChanged.onNext(phone)
                    }

                    phoneUpdated.onNext(phone)
                })
    }

    fun onGroupNotFound(disposableGroup: DisposableGroup? = null, callback: (Unit) -> Unit) = onChangeCallback(groupNotFound, callback, disposableGroup)
    fun onGroupChanged(disposableGroup: DisposableGroup? = null, callback: (Group) -> Unit) = onChangeCallback(groupChanged, callback, disposableGroup)
    fun onGroupUpdated(disposableGroup: DisposableGroup? = null, callback: (Group) -> Unit) = onChangeCallback(groupUpdated, callback, disposableGroup)
    fun onEventChanged(disposableGroup: DisposableGroup? = null, callback: (Event) -> Unit) = onChangeCallback(eventChanged, callback, disposableGroup)
    fun onPhoneChanged(disposableGroup: DisposableGroup? = null, callback: (Phone) -> Unit) = onChangeCallback(phoneChanged, callback, disposableGroup)
    fun onPhoneUpdated(disposableGroup: DisposableGroup? = null, callback: (Phone) -> Unit) = onChangeCallback(phoneUpdated, callback, disposableGroup)
    fun onContactInfoChanged(disposableGroup: DisposableGroup? = null, callback: (ContactInfo) -> Unit) = onChangeCallback(contactInfoChanged, callback, disposableGroup)
    fun onGroupMemberChanged(disposableGroup: DisposableGroup? = null, callback: (GroupMember) -> Unit) = onChangeCallback(groupMemberChanged, callback, disposableGroup)

    private fun <T> onChangeCallback(observable: Observable<T>, callback: (T) -> Unit, disposableGroup: DisposableGroup? = null) {
        (disposableGroup ?: on<DisposableHandler>().self()).add(observable.observeOn(AndroidSchedulers.mainThread()).subscribe({ callback.invoke(it) }, connectionError))
    }
}

class ContactInfo {
    val contactNames = mutableListOf<String>()
    val contactInvites = mutableListOf<String>()
}